File: surf-chat/src/main/java/dev/slne/surf/social/chat/service/ChatFilterService.java
package dev.slne.surf.social.chat.service;

import dev.slne.surf.social.chat.SurfChat;

import it.unimi.dsi.fastutil.objects.Object2ObjectMap;
import it.unimi.dsi.fastutil.objects.Object2ObjectOpenHashMap;
import it.unimi.dsi.fastutil.objects.ObjectArraySet;
import it.unimi.dsi.fastutil.objects.ObjectSet;

import java.io.BufferedReader;
import java.io.FileReader;

import lombok.Getter;
import lombok.extern.slf4j.Slf4j;

import net.kyori.adventure.text.Component;
import net.kyori.adventure.text.format.NamedTextColor;
import net.kyori.adventure.text.logger.slf4j.ComponentLogger;
import net.kyori.adventure.text.serializer.plain.PlainTextComponentSerializer;

import java.io.File;
import java.io.IOException;
import java.util.UUID;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.TimeUnit;
import java.util.regex.Pattern;

@SuppressWarnings("ResultOfMethodCallIgnored")
@Slf4j
@Getter
public class ChatFilterService {
  @Getter
  private static final ChatFilterService instance = new ChatFilterService();
  private final ObjectSet<String> blockedWords = new ObjectArraySet<>();
  private final ObjectSet<String> allowedDomains = new ObjectArraySet<>();
  private final ObjectSet<Pattern> blockedPatterns = new ObjectArraySet<>();

  private final ComponentLogger logger = ComponentLogger.logger(this.getClass());
  private final ConcurrentHashMap<UUID, Long> rateLimit = new ConcurrentHashMap<>();
  private final ConcurrentHashMap<UUID, Integer> messageCount = new ConcurrentHashMap<>();

  private static final Pattern VALID_CHARACTERS_PATTERN = Pattern.compile("^[a-zA-Z0-9/.:_,()%&()=?!<>|#^\"²³+*~-äöü@]*$");
  private static final long TIME_FRAME = TimeUnit.SECONDS.toMillis(10);
  private static final int MESSAGE_LIMIT = 5;

  public void loadBlockedWords() {
      File file = new File(SurfChat.getInstance().getDataFolder(), "blocked.txt");
      long start = System.currentTimeMillis();

      if (!file.exists()) {
        file.getParentFile().mkdirs();
        SurfChat.getInstance().saveResource("blocked.txt", false);
      }

      this.blockedWords.clear();
      this.blockedPatterns.clear();

      try (BufferedReader reader = new BufferedReader(new FileReader(file))) {
          String line;

          while ((line = reader.readLine()) != null) {
              String word = line.trim();
              if (!word.isEmpty()) {
                  this.blockedWords.add(word);

                String regex = this.getRegex(word);
                blockedPatterns.add(Pattern.compile(regex, Pattern.CASE_INSENSITIVE));
              }
          }
      } catch (IOException e) {
          logger.error("Failed to read blocked.txt file", e);
      }

      logger.info(Component.text("Loaded ", NamedTextColor.GREEN)
          .append(Component.text(blockedWords.size(), NamedTextColor.GOLD))
          .append(Component.text(" blocked words and their regexes in ", NamedTextColor.GREEN))
          .append(Component.text(System.currentTimeMillis() - start, NamedTextColor.GOLD))
          .append(Component.text("ms", NamedTextColor.GREEN)));
  }

  private String getRegex(String word) {
    Object2ObjectMap<Character, String> replacements = new Object2ObjectOpenHashMap<>();
    replacements.put('a', "[a@4]");
    replacements.put('b', "[b8]");
    replacements.put('c', "c");
    replacements.put('d', "d");
    replacements.put('e', "[e3]");
    replacements.put('f', "f");
    replacements.put('g', "[g9]");
    replacements.put('h', "h");
    replacements.put('i', "[i1!]");
    replacements.put('j', "j");
    replacements.put('k', "k");
    replacements.put('l', "[l1]");
    replacements.put('m', "m");
    replacements.put('n', "n");
    replacements.put('o', "[o0]");
    replacements.put('p', "p");
    replacements.put('q', "q");
    replacements.put('r', "r");
    replacements.put('s', "[s5]");
    replacements.put('t', "[t7]");
    replacements.put('u', "u");
    replacements.put('v', "v");
    replacements.put('w', "w");
    replacements.put('x', "x");
    replacements.put('y', "y");
    replacements.put('z', "[z2]");

    StringBuilder regexBuilder = new StringBuilder();
    for (char c : word.toCharArray()) {
        regexBuilder.append(replacements.getOrDefault(c, c + ""));
    }

    String regex = regexBuilder.toString();
    return regex;
  }

  public boolean containsBlocked(Component message) {
    return blockedPatterns.stream().anyMatch(pattern -> pattern.matcher(PlainTextComponentSerializer.plainText().serialize(message)).find());
  }

  public boolean containsLink(Component message) {
    String plainMessage = PlainTextComponentSerializer.plainText().serialize(message);
    String urlPattern = "((http|https|ftp)://)?([\\w-]+\\.)+[\\w-]+(/[\\w- ./?%&=]*)?";
    Pattern pattern = Pattern.compile(urlPattern, Pattern.CASE_INSENSITIVE);
    var matcher = pattern.matcher(plainMessage);

    while (matcher.find()) {
      String domain = matcher.group(3);
      if (allowedDomains.stream().noneMatch(domain::endsWith)) {
        return true;
      }
    }
    return false;
  }

  public boolean isValidInput(String input) {
    return VALID_CHARACTERS_PATTERN.matcher(input).matches();
  }

  public boolean isSpamming(UUID uuid) {
    long currentTime = System.currentTimeMillis();

    rateLimit.putIfAbsent(uuid, currentTime);
    messageCount.putIfAbsent(uuid, 0);

    long lastMessageTime = rateLimit.get(uuid);
    int count = messageCount.get(uuid);

    if (currentTime - lastMessageTime < TIME_FRAME) {
      if (count >= MESSAGE_LIMIT) {
        return true;
      } else {
        messageCount.put(uuid, count + 1);
        return false;
      }
    } else {
      rateLimit.put(uuid, currentTime);
      messageCount.put(uuid, 1);
      return false;
    }
  }
}

File: surf-chat/src/main/java/dev/slne/surf/social/chat/service/ChatFilterService.kt
package dev.slne.surf.social.chat.service

import dev.slne.surf.social.chat.SurfChat
import dev.slne.surf.surfapi.core.api.util.*
import net.kyori.adventure.text.Component
import net.kyori.adventure.text.serializer.plain.PlainTextComponentSerializer
import java.util.*
import kotlin.io.path.*
import kotlin.system.measureTimeMillis
import kotlin.time.Duration.Companion.seconds


private const val MESSAGE_LIMIT = 5

object ChatFilterService {

    private val log = logger()
    private val blockedWords = mutableObjectSetOf<String>()
    private val allowedDomains = mutableObjectSetOf<String>()

    private val blockedPatterns = mutableObjectSetOf<Regex>()
    private val rateLimit =
        mutableObject2LongMapOf<UUID>().apply { defaultReturnValue(0) }.synchronize()

    private val messageCount =
        mutableObject2IntMapOf<UUID>().apply { defaultReturnValue(0) }.synchronize()

    private val validCharactersRegex = "^[a-zA-Z0-9/.:_,()%&=?!<>|#^\"²³+*~-äöü@ ]*$".toRegex()
    private val urlRegex =
        "((http|https|ftp)://)?([\\w-]+\\.)+[\\w-]+(/[\\w- ./?%&=]*)?".toRegex(RegexOption.IGNORE_CASE)

    private val TIME_FRAME = 10.seconds.inWholeMilliseconds

    private val regexReplacements = char2ObjectMapOf(
        'a' to "[a@4]", 'b' to "[b8]", 'c' to "c", 'd' to "d",
        'e' to "[e3]", 'f' to "f", 'g' to "[g9]", 'h' to "h",
        'i' to "[i1!]", 'j' to "j", 'k' to "k", 'l' to "[l1]",
        'm' to "m", 'n' to "n", 'o' to "[o0]", 'p' to "p",
        'q' to "q", 'r' to "r", 's' to "[s5]", 't' to "[t7]",
        'u' to "u", 'v' to "v", 'w' to "w", 'x' to "x",
        'y' to "y", 'z' to "[z2]"
    )


    fun loadBlockedWords() {
        val duration = measureTimeMillis {
            val path = SurfChat.instance.dataPath / "blocked.txt"
            with(path) {
                createDirectories()
                if (!exists()) createFile()
            }

            blockedWords.clear()
            blockedPatterns.clear()

            path.useLines { lines ->
                lines.map { it.trim() }
                    .filter { it.isNotEmpty() }
                    .forEach {
                        blockedWords.add(it)

                        val regex = getRegex(it)
                        blockedPatterns.add(Regex(regex, RegexOption.IGNORE_CASE))
                    }
            }
        }

        log.atInfo()
            .log("Loaded %s blocked words and their regexes in %sms", blockedWords.size, duration)
    }


    private fun getRegex(word: String): String {
        val regexBuilder = StringBuilder(word.length)
        for (c in word.toCharArray()) {
            regexBuilder.append(regexReplacements.getOrDefault(c, c.toString()))
        }

        return regexBuilder.toString()
    }

    fun containsBlocked(message: Component) = blockedPatterns.any {
        it.containsMatchIn(
            PlainTextComponentSerializer.plainText().serialize(message)
        )
    }


    fun containsLink(message: Component): Boolean {
        val plainMessage = PlainTextComponentSerializer.plainText().serialize(message)

        return urlRegex.findAll(plainMessage).any { result ->
            val domain = result.groupValues.getOrNull(3) ?: return@any false
            allowedDomains.none { domain.endsWith(it) }
        }
    }

    fun isValidInput(input: String): Boolean {
        return validCharactersRegex.matches(input)
    }

    fun isSpamming(uuid: UUID): Boolean {
        val currentTime = System.currentTimeMillis()
        val lastMessageTime = rateLimit.getLong(uuid)
        val count = messageCount.getInt(uuid)

        return if (currentTime - lastMessageTime < TIME_FRAME) {
            (count >= MESSAGE_LIMIT).also { if (!it) messageCount[uuid] = count + 1 }
        } else {
            rateLimit[uuid] = currentTime
            messageCount[uuid] = 1
            false
        }
    }
}